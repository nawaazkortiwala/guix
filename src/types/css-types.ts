/**
 * @file This file contains all the types related to the global CSS Properties available for all core building blocks.
 * @author Nawaaz Kortiwala <nawaazkortiwala@gmail.com>
 *
 * This file is part of the MyProject open source project, which can be found at https://github.com/nawaazkortiwala/guix.
 *
 * Copyright (c) 2023 Nawaaz Kortiwala
 *
 * Licensed under the Apache 2.0 License. See LICENSE.txt for more information.
 */

import { StandardProperties } from 'csstype'

export type Theme = object

export type Context = object

export type CSSValue<T extends keyof StandardProperties> = StandardProperties[T]

export type CSSValueFn<T extends keyof StandardProperties> = (
  theme: Theme,
  context: Context
) => CSSValue<T>

export type CSSProperties = {
  [key in keyof StandardProperties as `$${string & key}`]: CSSValue<key> | CSSValueFn<key>
}
