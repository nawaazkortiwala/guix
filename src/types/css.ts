/**
 * Contains all the types related to CSS Properties.
 */
export declare type CSSProperties = {
  [key: string]: string | number | CSSProperties;
};
