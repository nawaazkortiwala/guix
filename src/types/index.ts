export * from './css-types'
export * from './element-types'
export * from './state-types'
