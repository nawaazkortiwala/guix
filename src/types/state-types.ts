export type ThemeContent = {
  baseSize: number // px
  colors: {
    primary: string
    secondary: string
    tertiary: string
    success: string
    warning: string
    danger: string
    info: string
    light: string
    dark: string
    white: string
    black: string
    custom: Record<string, string>
  }
  fonts: {
    primary: string
    secondary: string
    tertiary: string
    custom: Record<string, string>
  }
  fontWeights: {
    extraLight: number
    light: number
    normal: number
    semiBold: number
    bold: number
    extraBold: number
    custom: Record<string, number>
  }
  lineHeights: {
    xs: number
    sm: number
    md: number
    lg: number
    xl: number
    x2l: number
    x3l: number
    custom: Record<string, number>
  }
  letterSpacings: {
    xs: number
    sm: number
    md: number
    lg: number
    xl: number
    x2l: number
    x3l: number
    custom: Record<string, number>
  }
  space: {
    xs: number
    sm: number
    md: number
    lg: number
    xl: number
    x2l: number
    x3l: number
    custom: Record<string, number>
  }
  radii: {
    xs: number
    sm: number
    md: number
    lg: number
    xl: number
    x2l: number
    x3l: number
    custom: Record<string, number>
  }
  shadows: {
    xs: string
    sm: string
    md: string
    lg: string
    xl: string
    x2l: string
    x3l: string
    custom: Record<string, string>
  }
  zIndices: {
    xs: number
    sm: number
    md: number
    lg: number
    xl: number
    x2l: number
    x3l: number
    custom: Record<string, number>
  }
  breakpoints: {
    xs: number
    sm: number
    md: number
    lg: number
    xl: number
    x2l: number
    x3l: number
    custom: Record<string, number>
  }
  mediaQueries: {
    xs: string
    sm: string
    md: string
    lg: string
    xl: string
    x2l: string
    x3l: string
    custom: Record<string, string>
  }
  custom: Record<string, unknown>
}

export type ThemeMode = string

export type Theme = Record<ThemeMode, ThemeContent>

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Context = {
  theme: Theme
  activeTheme: ThemeMode
  themeModes: ThemeMode[]
}
