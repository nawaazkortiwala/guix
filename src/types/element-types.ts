import React, { HTMLAttributes } from 'react'
import { IStyledComponent } from 'styled-components'

import { CSSProperties, CSSStyleFn } from './css-types'

type FastOmit<T extends object, U extends string | number | symbol> = {
  [K in keyof T as K extends U ? never : K]: T[K]
}

type Substitute<A extends object, B extends object> = FastOmit<A, keyof B> & B

interface ReactAttributes<T> extends HTMLAttributes<T> {}

export type ElementProps<T> = React.HTMLAttributes<T> & CSSProperties<T>

export type ElementPropsWithChildren<T> = React.PropsWithChildren<ElementProps<T>>

export type Element<R extends ReactAttributes<E>, E> = IStyledComponent<
  'web',
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Substitute<React.DetailedHTMLProps<R, E>, CSSProperties<any>>
>

// eslint-disable-next-line @typescript-eslint/ban-types
export type StyleFnObject<T = {}> = Record<string | number | symbol, CSSStyleFn<T>>
