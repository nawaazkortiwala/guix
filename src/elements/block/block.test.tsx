import { CSSProperties } from '@/types/css-types'
import '@testing-library/jest-dom'
import 'jest-styled-components'
import React from 'react'
import renderer from 'react-test-renderer'
import Block from './block'

describe('Block has given styles', () => {
  it('should render with specific CSS properties', () => {
    const props: CSSProperties = {
      $alignItems: 'center',
      $backgroundColor: 'red',
      $color: 'white',
    }
    const tree = renderer.create(<Block data-testid="block" {...props} />).toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree).toHaveStyleRule('align-items', 'center')
    expect(tree).toHaveStyleRule('background-color', 'red')
    expect(tree).toHaveStyleRule('color', 'white')
  })

  it('should match snapshot', () => {
    const tree = renderer.create(<Block />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
