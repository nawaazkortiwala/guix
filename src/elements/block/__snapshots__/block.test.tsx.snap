// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block has given styles should match snapshot 1`] = `
<div
  className=""
/>
`;

exports[`Block has given styles should render with specific CSS properties 1`] = `
.c0 {
  align-items: center;
  background-color: red;
  color: white;
}

<div
  className="c0"
  data-testid="block"
/>
`;
